### DISCUSSION TITLE
Issue with QC pipeline step

### DISCUSSION THREAD

**Discussion by marcellamontagnese** [McIntosh-Lab/tvb-ukbb#186] [Help]
Hello, 

Thanks a lot for this resource! I have been running a slighly modified version of the pipeline on CPU and I am encountering issues with the QC pipeline step. I keep getting the error:

` File "/opt/fsl/fslpython/envs/fslpython/lib/python3.7/site-packages/OpenGL/raw/osmesa/mesa.py", line 10, in _f
    function,_p.PLATFORM.OSMesa,
AttributeError: 'GLXPlatform' object has no attribute 'OSMesa'`

See attached the log for further info.

Thanks in advance for any help!
[tvb_bb_QC_sub-33909.log](https://github.com/McIntosh-Lab/tvb-ukbb/files/11422782/tvb_bb_QC_sub-33909.log)

**Comment by justinliuyuwang** [comment]
Hi Marcella!

Great to see you trying out the pipeline. 

1. Are you using the Singularity container? 
2. What operating system are you using? 
3. Could you please share the entire log folder?

Thanks,
Justin

**Reply by marcellamontagnese** [reply]
Hi Justin, thanks for the quick reply!

Regarding your questions, I am using the Singularity container on a High Performance Cluster running a Linux operating system with kernel version 3.10.0-1160.88.1.el7.x86_64. 

Please see attached the entire log folder (at the moment there is an issue with the SWI file manager function that I still need to fix + I made some preliminary edits to the diffusion pipeline given that we have a full set of data in the reverse phase encoding direction)

Thank you, 
Marcella 
[logs.zip](https://github.com/McIntosh-Lab/tvb-ukbb/files/11423619/logs.zip)

**Reply by justinliuyuwang** [reply]
Hi Marcella!

Thank you for your patience with this. We're still looking into the OSMesa issue.

I'm not sure if this is something you are currently addressing, but you seem to have some issues with your pipeline's probtrackx2 tractography calls. As a result, the pipeline currently doesn't seem to be creating structural connectivity outputs.

It looks like you're using the main branch of the pipeline. We will be updating the main branch this weekend, but in the meantime, I would recommend running your data with the ADNI3 branch of the pipeline.

Also, I highly recommend clearing the contents of your subject folder in between runs (i.e. erase everything except the rawdata directory and its contents or use a fresh copy of the subject if you want to run the subject through the pipeline again).

**Reply by justinliuyuwang** [reply]
Before the next time you run the pipeline, could you please try putting the following line:
`export PYOPENGL_PLATFORM=osmesa` before the  `source activate $SOFTDIR/software/env` line in `init_vars`?

Potential solution here https://github.com/mcfletch/pyopengl/issues/10#issuecomment-514959591 and in the comment below the linked comment.

**Reply by justinliuyuwang** [reply]
Hi Marcella!

Thanks to @yilewang, I believe we may have a solution to this issue now. Please try using or adapting from the new `Cam-CAN_dev_CPU` or `ADNI3_dev_CPU` branches. Modified instructions for the pipeline setup can be found [here](https://github.com/McIntosh-Lab/tvb-ukbb/wiki/FAQ#my-qc-pipeline-isnt-running-correctly-i-am-getting-a-attributeerror-glxplatform-object-has-no-attribute-osmesa-error). 

Please note that this updated version of the pipeline (and container) will take longer to process subjects than prior GPU versions did. Updated GPU versions of the pipeline and container are currently in the works.

Hopefully this is still helpful! Thanks again for bringing up this issue.

Best,
Justin

**Reply by marcellamontagnese** [reply]
Hi Justin, 

Apologies I thought I had replied to your previous comment! Using your previous suggestions had helped fix the osmesa issue.
And thanks for this new update - I will make sure to try using the new `Cam-CAN_dev_CPU` or `ADNI3_dev_CPU` branches.

Thanks again for the assistance with the issues I encountered.

Best, 
Marcella

**Reply by justinliuyuwang** [reply]
Thank you for your patience! Please note that the tract lengths produced by those two branches may not be completely accurate at the moment - we are working on fixing that right now.
